keras_retinanet/utils/plugin.py

```diff
@@ -0,0 +1,33 @@
+"""
+Copyright 2017-2018 Ashley Williamson (https://inp.io)
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+from yapsy.IPlugin import IPlugin
+
+
+class DatasetPlugin(IPlugin):
+    def __init__(self):
+        self.dataset_type = None
+        super(DatasetPlugin, self).__init__()
+
+
+    def parser_args(self, parser):
+        pass
+
+    def check_args(self, parsed_args):
+        pass
+
+    def get_generator(self, args):
```

Comment:
Here  raising a `NotImplementedError` definitely seems like the best default implementation.

Regarding the name of the function, `get_...` to me implies a cheap retrieval operation. A dataset might do arbitrary things when being constructed, including reading files from disk or even the internet. In general, it's not expected to be a cheap operation, so I would prefer a name like `make_generators` or `create_generators`. In it's current form the name should also be plural since the function can return multiple generators.

I'm a bit on the fence if we should return a dictionary with multiple generators, or have separate functions for the train and evaluation generator (which might return None). Or maybe we should just return one generator and invoke it twice if we also want an evaluation generator. This last option sounds best but would probably require a change in the way we parse arguments.

Reply:
Yes! `NotImplementedError`. Agreed.

Replacing `get_` with a more 'cost associative word' makes sense. Not sure that `create_` fits either. It is a get, as it returns a generator. Yet create might not imply a return.

Not quite sure I follow what you're suggesting for return functions ( wrapping the generator in a lambda??? ), nor the invoking twice eval generator. Can I get some clarification on that?

Reply:
> Replacing get_ with a more 'cost associative word' makes sense. Not sure that create_ fits either. It is a get, as it returns a generator. Yet create might not imply a return.

To me `create` and `make` do imply a return. But if you have an alternative that fits, that also works for me.

> Not quite sure I follow what you're suggesting for return functions ( wrapping the generator in a lambda??? ), nor the invoking twice eval generator. Can I get some clarification on that?

What I mean is that now we have one function that has to return two generators depending on the arguments: a generator for training and one for evaluation. That means each dataset plugin now has to checks if the user wants an evaluation generator.

We could also have two functions instead of one function returning two generators. Then a global option can disable evaluation without specific support from the dataset plugins.

Even neater would be one function that can be called multiple times to create multiple generators with different parameters. That would be a simpler interface for the dataset plugin and allow more flexibility in using it. However, that would require being able to divide arguments into separate calls to `make_generator`, and currently our argument parsing isn't capable of doing that.

Reply:
We could explicitly have `create_training_generator` and `create_evaluation_generator`, is this what you mean? We can then catch `NotImplemented` exceptions within the test script if a Dataset doesn't provide the means to train or evaluate?
Typically, the difference between Evaluation and Training generators is just one or two baked in args. The majority of the args passed to that function are used by both anyhow, which then makes sense to return both generators.

Reply:
> We could explicitly have create_training_generator and create_evaluation_generator, is this what you mean?

That is one option I was considering, yes.

> We can then catch NotImplemented exceptions within the test script if a Dataset doesn't provide the means to train or evaluate?

NotImplemented is not allowed for the training generator, so wouldn't be caught for that. As for `create_evaluation_generator`, that could simply have a default implementation that returns `None`.

I do prefer splitting because it removes the reponsibility of checking if `--no-evaluation` was passed from each dataset plugin. Suppose that we want to rename that flag in the future, that would be pretty difficult if each dataset plugin has to check it.

Ideally, I don't even want plugins to have access to global options, but right now that might complicate things too far. But at least splitting the functions is a step in the right direction.

Reply:
I agree it should be the script responsibility for whether to actually use `evaluation`. It still is entirely possible to have a Dataset which isn't meant to be trained on, just provides an Evaluation mechanism. Think transfer learning without finetuning for example. I'm thinking beyond train.py a little bit here. Dataset just provides what it can provide in those contexts.

Base plugin will have NotImplError for those functions. Datasets should override those and either return a `Generator` or `None`. Any script calling to `create_evaluation_generator` or `create_training_generator` should expect to check `if not None` prior to utilising. And even then we can wrap the use of a generator in a `try` block to catch anything unexpected.

Reply:
> I agree it should be the script responsibility for whether to actually use evaluation. It still is entirely possible to have a Dataset which isn't meant to be trained on, just provides an Evaluation mechanism.

True. Treating the train generator specially doesn't really make sense from the plugins point of view.

> Base plugin will have NotImplError for those functions. Datasets should override those and either return a Generator or None. Any script calling to create_evaluation_generator or create_training_generator should expect to check if not None prior to utilising. And even then we can wrap the use of a generator in a try block to catch anything unexpected.

We *could* also make them all return `None` in the base interface I suppose. Or document that if a dataset plugin doesn't support one type of dataset generator that it should leave it raising a `NotImplementedError`. Otherwise each script that wants to use a specific generator needs to check for both `None` and `NotImplementedError`.